Small fixes and get_mindat_json refactor.

CorylC opened the pull request:
- standardized feedback for bad string arguments in geomaterial.py functions
- Added fix for longer downloads getting a JSONDecodeError
- Added accommodation for GeoRegionRetriever
- fixed the expand function in geomaterials.py
- refactored varieties in geomaterialIdRetriever
- Changed expand in LocalitiesRetriever to match GeomaterialsRetriever

CorylC committed:
first fix of locgeoregion2

Added a special case for downloading locgeoregion2.

CorylC committed:
Large file download fix.

Added a catch for jsonDecodeErrors where it waits to let the system reload and then tries again.

CorylC committed:
Expand issue

Expand fields does not work with lists well, so I changed the lists to a string and changed the harmful * character to be "*" in the param.

CorylC committed:
Fixes

-Standardized valid option types in string based parameters in geomaterials.py
-Refactored varieties in geomaterialIdRetriever.
-Changed expand to match geomaterials in localitiesRetriever()
-

CorylC committed:
Refactored get_mindat_json

Changed get_mindat_json to be a little bit cleaner and work better for the locgeoregion2 case.

CorylC committed:
Removed comment

Removed commented out function.

ChuBL commented on openmindat/geomaterials.py:
I appreciate your effort on input validations. However, I have some comments. Except for type issues, I suggest we do not use hard input validation with a list of valid options but only print a message indicating that the inputs might be wrong and attaching with a corresponding redoc link. You may refer to my modification on the `country` function of `localities.py` for references. This modification will allow users to input "invalid" params because the allowed input might change according to API developments. The hard-coded restriction is not a good idea for a continuously developing API.

CorylC replied:
I have changed all of the raise error statements to print statements to make them suggestions instead of hard input validation.

ChuBL commented on openmindat/localities.py:
```diff
@@ -191,9 +193,12 @@ def expand(self, EXPAND_FIELDS):
 
-        expand_fields = EXPAND_FIELDS
+                #Changes * to "*" and from a list to a string since this is what the param will read without failing.
+        expand_fields = [w.replace('*', '"*"') for w in expand_fields]
```
I am wondering if the initial user input is a `"*"` and your code replaces it with `""*""`. Will the API server still recognize and process it? Again, do not `raise Value Error` for the "invalid" inputs, but printing it as a soft warning would be better.

ChuBL commented on openmindat/mindat_api.py:
Your modification seems reasonable. Since this function has a broad impact on other classes, I assume you have tested through most of the cases and proven the modification to be functional. Otherwise, please revise it until it passes the tests.

CorylC replied:
I have just finished testing every endpoint that works and it all seems to be working well.

ChuBL reviewed (changes requested):
Thank you for your updates! Please take a look at the comments and make the necessary improvements. Thanks.

CorylC commented:
I will go through and change it from error messages to print statements to make them suggestions. I have done the testing for * and "​*", both work. And for the api changes I have tested endpoints that usually gives some trouble as well as others through normal testing, I will go through and test all of them officially but so far I haven't had any issues.
________________________________
From: Jiyin Zhang ***@***.***>
Sent: Thursday, May 16, 2024 3:23 PM
To: ChuBL/OpenMindat ***@***.***>
Cc: CorylC ***@***.***>; Author ***@***.***>
Subject: Re: [ChuBL/OpenMindat] Small fixes and get_mindat_json refactor. (PR #15)


@ChuBL requested changes on this pull request.

Thank you for your updates! Please take a look at the comments and make the necessary improvements. Thanks.

________________________________

On openmindat/geomaterials.py<https://github.com/ChuBL/OpenMindat/pull/15#discussion_r1604118365>:

I appreciate your effort on input validations. However, I have some comments. Except for type issues, I suggest we do not use hard input validation with a list of valid options but only print a message indicating that the inputs might be wrong and attaching with a corresponding redoc link. You may refer to my modification on the country function of localities.py for references. This modification will allow users to input "invalid" params because the allowed input might change according to API developments. The hard-coded restriction is not a good idea for a continuously developing API.

________________________________

In openmindat/localities.py<https://github.com/ChuBL/OpenMindat/pull/15#discussion_r1604121431>:

> @@ -191,9 +193,12 @@ def expand(self, EXPAND_FIELDS):
         if invalid_options:
             raise ValueError(f"Invalid EXPAND_FIELDS: {', '.join(invalid_options)}\nEXPAND_FIELDS must be one or more of the following: {', '.join(valid_options)}")

-        expand_fields = EXPAND_FIELDS
+                #Changes * to "*" and from a list to a string since this is what the param will read without failing.
+        expand_fields = [w.replace('*', '"*"') for w in expand_fields]


I am wondering if the initial user input is a "*" and your code replaces it with ""*"". Will the API server still recognize and process it? Again, do not raise Value Error for the "invalid" inputs, but printing it as a soft warning would be better.

________________________________

On openmindat/mindat_api.py<https://github.com/ChuBL/OpenMindat/pull/15#discussion_r1604126739>:

Your modification seems reasonable. Since this function has a broad impact on other classes, I assume you have tested through most of the cases and proven the modification to be functional. Otherwise, please revise it until it passes the tests.

—
Reply to this email directly, view it on GitHub<https://github.com/ChuBL/OpenMindat/pull/15#pullrequestreview-2061999929>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BCW4I3N7PCPP7LZY6RQNKWLZCUWXVAVCNFSM6AAAAABH3AP2L2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANRRHE4TSOJSHE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

CorylC committed:
Changed to warning messages.

Changed from hard input validation to warnings.

CorylC committed:
Small fixes

filtered out none so it's easier to use join on invalid input list.
changed minerals_ima to not have hard input validation.

ChuBL reviewed (approved):
Great job. However, instead of printing the validation objects, I would provide a link to the redoc's section so users can refer to the up-to-date valid options.
